py/server/deephaven/table_reader.py

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+#
+"""This module supports reading the data in a Deephaven table in a chunked manner."""
+
+from typing import Union, Sequence, Generator, Dict, Optional
+
+import jpy
+import numpy as np
+
+from deephaven.column import Column
+from deephaven.jcompat import to_sequence
+from deephaven.numpy import _column_to_numpy_array
+from deephaven.table import Table
+
+_JTableUpdateDataReader = jpy.get_type("io.deephaven.integrations.python.PythonTableDataReader")
+
+
+def _col_defs(table: Table, cols: Union[str, Sequence[str]]) -> Sequence[Column]:
+    if not cols:
+        col_defs = table.columns
+    else:
+        cols = to_sequence(cols)
+        col_defs = [col for col in table.columns if col.name in cols]
+
+    return col_defs
+
+def table_reader(table: Table, cols: Optional[Union[str, Sequence[str]]] = None, *, chunk_size: Optional[int] = None) \
```

Comment:
Let me be explicit here.  The current syntax is:
1. inconsistent with the listener syntax
2. awkward if you just want to pull the whole table.

I think:
1. `table_reader` should be renamed to `table_reader_chunks`.
2. There should be a new method `table_reader` that reads everything into a dict all at once.  No generator.  This would be analogous to the UI for `TableUpdate.added()` or `TableUpdated.removed()`.  This would simplify the syntax for the all-at-once case.

Reply:
We shouldn't complicate the interface with a distinction here. We have perfectly good methods for pulling the entire table... `to_numpy`, `to_pandas`, and `to_arrow`, and they are more efficient in terms of locking. This task is to add an iterator, not to add yet another way to pull the entire table. The listener interface has the distinct use-case of being over a sub-set of rows, and being inside a listener.

The only thing I think we should consider adding here is adding an (all Python) higher-level element-by-element iterator *on top* of the chunk-by-chunk iterator. It would be very inefficient, creating a dict per row, but at least the JNI crossing would be per chunk.

Reply:
If the locking is more efficient in those listed methods, I think I would be ok leaving this as one method if the pydocs are extended to tell users that those methods are more efficient if the whole table is being pulled.

The row iterator may be a good addition.  Pandas has something similar.  See Method 4 in https://www.geeksforgeeks.org/different-ways-to-iterate-over-rows-in-pandas-dataframe/#.